src/Vim.re

```diff
@@ -33,7 +39,34 @@ let flushQueue = () => {
   queuedFunctions := [];
 };
 
-let checkAndUpdateState = f => {
+let synchronizeAndUpdateState = (~context: Context.t, f) => {
```

Comment:
Isn't this really just `updateState?. "Synchronizing" and "updating" seems to be essentially the same thing here.

Reply:
It's synchronizing the `context` passed in via `~context`.

Reply:
I still don't understand what the synchronization here is, as a process separate from updating the state.

Reply:
It's really doing 3 things:
- Taking the `context` passed in, and making sure the state of the world matches. So if we pass in a context with a bufferId of `2`, that will make sure that buffer `2` is the active buffer. Likewise for auto-closing pairs, comment settings, indentation, etc.
- Running a command with side-effects, and watching those side-effects (so they are returned as part of the output)
- Collect the `context` after the command - the state of the world - and return it.

So I guess it could be something like `synchronizeAndWatchEffectsAndReturnNewContext` but might be a bit verbose. Maybe just `synchronizeAndWatchEffects`?

Reply:
Right, yeah that's a good sign that the function does way too much. Also the fact that it's over 100 lines long, which makes ti very difficult to get a good overview of what it's actually doing. So splitting them into several smaller functions and then composing them here would be a good idea I think.

As for the name, I think just `runWith(~context, () => ...)` would work? Then if it's descriptive enough on its own, because you've split it into several smaller functions, you shouldn't even need to explain it with a comment.

Reply:
Sounds good, I went with `runWith` (and because I pulled the `EffectWatcher` out, so I can at least get the `Context.t` changes, there no longer is the watching-side-effects piece)